File: seat.py
import pandas as pd
import csv

# Function to assign guests to tables based on family_id
# Each table can accommodate 10 guests

def assign_guests_to_tables(csv_file):
    tables = []
    family_members = {}

    with open(csv_file, 'r') as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            name = row['name']
            family_id = int(row['family_id'])

            if family_id not in family_members:
                family_members[family_id] = []

            family_members[family_id].append(name)

    guests = []
    for family_id, members in family_members.items():
        guests.extend(members)

    tables = [guests[i:i+10] for i in range(0, len(guests), 10)]

    # Write seating assignments to guest_seat.csv
    with open('guest_seat.csv', 'w', newline='') as csvfile:
        fieldnames = ['name', 'table_number']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()
        for table_number, table in enumerate(tables, start=1):
            for guest in table:
                writer.writerow({'name': guest, 'table_number': table_number})

def main():     
    csv_file = 'guest_list.csv'
    assign_guests_to_tables(csv_file)
    
# Main function to run the assignment 
if __name__ == "__main__":
    main()  

File: seat.R
# Load necessary library
library(dplyr)

# Read the CSV file
guests <- read.csv('guest_lst.csv')

# Function to assign seats based on family_id and handle single guests
assign_seats <- function(guests) {
  # Sort guests by family_id (if not NULL) and name to ensure families are seated together
  sorted_guests <- guests %>%
    arrange(ifelse(is.na(family_id), Inf, family_id), name)
  
  # Initialize seat assignment
  seat_assignment <- data.frame(
    name = sorted_guests$name,
    family_id = sorted_guests$family_id,
    table_number = NA,
    seat_number = NA
  )
  
  # Assign seats
  table_count <- 1
  seat_count <- 1
  
  for (i in 1:nrow(sorted_guests)) {
    if (seat_count > 10) {
      table_count <- table_count + 1
      seat_count <- 1
    }
    
    seat_assignment$table_number[i] <- table_count
    seat_assignment$seat_number[i] <- seat_count
    
    seat_count <- seat_count + 1
  }
  
  return(seat_assignment)
}

# Assign seats to guests
seating_arrangement <- assign_seats(guests)

# Write the seating arrangement to a CSV file
write.csv(seating_arrangement, 'guest_seat.csv', row.names = FALSE)
